python/cudf/cudf/core/scalar.py

```diff
@@ -16,7 +17,50 @@
 )
 
 
-class Scalar(BinaryOperand):
+# Note that the metaclass below can easily be generalized for use with
+# other classes, if needed in the future. Simply replace the arguments
+# of the `__call__` method with `*args` and `**kwargs`. This will
+# result in additional overhead when constructing the cache key, as
+# unpacking *args and **kwargs is not cheap. See the discussion in GH
+# #11246 for more details.
+class CachedScalarInstanceMeta(type):
+    """
+    Metaclass for Scalar that caches `maxsize` instances.
+    """
+
+    def __new__(self, names, bases, attrs, maxsize=128):
+        self.__maxsize = maxsize
+        return type.__new__(self, names, bases, attrs)
+
+    def __init__(self, names, bases, attrs, **kwargs):
+        self.__instances = OrderedDict()
+
+    def __call__(self, value, dtype=None):
+        # the cache key is constructed from the arguments, and also
+        # the _types_ of the arguments, since objects of different
+        # types can compare equal
+        arg_tuple = (value, type(value), dtype, type(dtype))
+        try:
+            # try retrieving an instance from the cache:
+            self.__instances.move_to_end(arg_tuple)
```

Comment:
We should document somewhere that this class implements caching with LRU semantics, otherwise the usage of `move_to_end` and `popitem` could be confusing.

Reply:
Doen